Combination

pdmosses opened the pull request:
This PR combines (and resolves conflicts between) #448, #463, #466, #494, #495, #496, #498, and #572. 

The main aim is to facilitate use of several of the implemented features _together_, when using the fork as a remote theme. It should also simplify merging the included PRs into a future release.

The branch [combination-rec-nav](https://github.com/pdmosses/just-the-docs/tree/combination-rec-nav) adds [multi-level navigation](https://github.com/pmarsceill/just-the-docs/pull/462) and (NEW:) [sibling links](https://github.com/pmarsceill/just-the-docs/pull/394) to the branch used for this PR. It includes updated [documentation for the navigation structure](https://pdmosses.github.io/just-the-docs/docs/navigation-structure/), and reorganised and extended [navigation tests](https://pdmosses.github.io/just-the-docs/tests/navigation/). The documentation and the tests can be browsed at the (temporary) [website published from the combination-rec-nav branch](https://pdmosses.github.io/just-the-docs/).

_Caveat:_ The changes to v0.3.3 in this PR and #462 have not yet been reviewed or approved, and may need updating before merging into a release of the theme. If you use a branch from a PR as a remote theme, there is a risk of such updates affecting your website. Moreover, these branches are likely to be deleted after they have been merged. To avoid  such problems, you could copy the branch that you want to use to your own fork of the theme.

pdmosses commented:
https://github.com/pmarsceill/just-the-docs/pull/578/commits/91cdd5b59aa731773536854c5e4c8546b74ccb90 stops `nav_fold: true` affecting the indentation and color of the links shown when the collection is unfolded.

It also ensures that collection names stay on the left of the navigation panel with narrower displays.

The same changes have been made and published in the [combination-rec-nav](https://pdmosses.github.io/just-the-docs/) branch (but not in the original `pages-and-folded-collections` branch used for PR #572).

pdmosses commented:
My [combination-rec-nav](https://github.com/pdmosses/just-the-docs/tree/combination-rec-nav) branch (extending this PR) now includes support for both manual and auto toggling between custom color schemes, with various [configuration options](https://pdmosses.github.io/just-the-docs/docs/configuration/#toggle-between-two-schemes).

jthrilly commented:
This work constitutes a huge improvement to the theme. @pmarsceill - is there anything that can be done to help with reviewing/merging?

odashevskii-plaid commented:
@pdmosses it looks like search doesn't work in https://pdmosses.github.io/just-the-docs/
I've tracked down the problem to this rule:

```css
html .search-active .search-results, html[data-theme="light"] .search-active .search-results {
    display: block
}
```
not being effective because the `search-active` class is on `html` tag itself; so it should be `html.search-active` ...

pdmosses commented:
@odashevskii-plaid many thanks for reporting the problem with search on https://pdmosses.github.io/just-the-docs/, and especially for tracking down its cause!

The website at https://pdmosses.github.io/just-the-docs/ is produced from https://github.com/pdmosses/just-the-docs/tree/combination-rec-nav, which extends the https://github.com/pdmosses/just-the-docs/tree/combination branch used for the present PR with further features, including a configurable toggle for dark mode. 

The cause of the reported problem appears to be an interaction between the CSS for the search and toggle features, so it doesn't affect this PR directly, and search works as usual in a local build of the branch in this PR. Search also works with  [combination-rec-nav](https://github.com/pdmosses/just-the-docs/tree/combination-rec-nav) when the toggle configuration is omitted in `_config.yml`.

The [relevant commit](https://github.com/pdmosses/just-the-docs/commit/f0408bdd15d662b9ad2f08871f3e201a9d7e2d46) to [combination-rec-nav](https://github.com/pdmosses/just-the-docs/tree/combination-rec-nav) uses the following SCSS in [_includes/css/just-the-docs-toggle.scss.liquid](https://github.com/pdmosses/just-the-docs/blob/combination-rec-nav/_includes/css/just-the-docs-toggle.scss.liquid):

```scss
html, html[data-theme="light"] {
  @import "./support/support";
  @import "./color_schemes/{{ include.color_scheme }}";
  @import "./modules";
  {% include css/custom.scss.liquid color_scheme = include.color_scheme %}
  {% include css/callouts.scss.liquid color_scheme = include.color_scheme %}
}

html[data-theme="dark"] {
  @import "./support/support";
  @import "./color_schemes/{{ include.toggle_color_scheme }}";
  @import "./modules";
  {% include css/custom.scss.liquid color_scheme = include.toggle_color_scheme %}
  {% include css/callouts.scss.liquid color_scheme = include.toggle_color_scheme %}
}
```

I copied that pattern from  https://derekkedziora.com/blog/dark-mode (naively hoping that it would work also without using CSS variables). I don't know much about SCSS, and it's unclear to me whether it's possible to adjust the above code to avoid the problem with the search feature. If not, perhaps the CSS used for search could be changed to avoid the use of `html.search-active`?

odashevskii-plaid commented:
@pdmosses I came up with a simple fix for this. If we attach `search-active` to the `body` tag, it works:

```
diff --git a/assets/js/just-the-docs.js b/assets/js/just-the-docs.js
index c287ec0..61e609d 100644
--- a/assets/js/just-the-docs.js
+++ b/assets/js/just-the-docs.js
@@ -74,7 +74,7 @@ function initSearch() {
   request.onload = function(){
     if (request.status >= 200 && request.status < 400) {
       var docs = JSON.parse(request.responseText);
-
+
       lunr.tokenizer.separator = {{ site.search.tokenizer_separator | default: site.search_tokenizer_separator | default: "/[\s\-/]+/" }}

       var index = lunr(function(){
@@ -120,12 +120,14 @@ function searchLoaded(index, docs) {
   var currentInput;
   var currentSearchIndex = 0;

+  var body = document.querySelector('body');
+
   function showSearch() {
-    document.documentElement.classList.add('search-active');
+    body.classList.add('search-active');
   }

   function hideSearch() {
-    document.documentElement.classList.remove('search-active');
+    body.classList.remove('search-active');
   }

   function update() {
```

pdmosses commented:
@odashevskii-plaid excellent! that does indeed fix the search feature, and I've applied the fix in my [combination-rec-nav ](https://github.com/pdmosses/just-the-docs/tree/combination-rec-nav) branch. Search is now working together with the toggle button at https://pdmosses.github.io/just-the-docs/

For the record, I failed to find a workaround by changing the SCSS used to support toggling: I tried replacing `html` by `body`, but that badly affected the overall styling. And trying to use the parent selector gave Sass errors.

pdmosses commented:
Added the nav-scroll feature from PR #639 to the `combination` branch.

Added the same feature to the `combination-rec-nav` branch (adjusted for use with `relative-url`). See the published site at https://pdmosses.github.io/just-the-docs/

rayward commented:
@pdmosses given that this repo doesn't seem to be maintained anymore and you've been very active, perhaps you could just merge all these fixes and features down into your master branch and start maintaining an active fork there where it can be contributed to?

AlanFoster commented:
I was able to replicate the crash observed in https://github.com/pmarsceill/just-the-docs/pull/462#issuecomment-914264400 with this newer branch too :+1:

Gemfile:
> gem "just-the-docs", github: "pdmosses/just-the-docs", branch: "combination-rec-nav"

Crash:
```
Liquid Exception: Liquid error (...): Nesting too deep
```

Just wanted to cross-reference it with this pull request, as I was sure where the latest work was being tracked! 🏆

AlanFoster commented:
Ran into a small liquid error when switching from the `rec-nav-2` branch to the `combination-rec-nav` branch:
> Liquid Exception: Liquid error (.../bundler/gems/just-the-docs-429b51844fbc/_includes/nav/main.html line 57): Cannot sort a null object. included in ...bundler/gems/just-the-docs-429b51844fbc/_layouts/default.html

To get around this problem I needed to update my `_config.yml` file with:
```yaml
just_the_docs:
  collections: []
```

Potentially unrelated to this PR, I also noticed there were multiple aux links appearing:

![image](https://user-images.githubusercontent.com/1271782/132729550-d9d833af-efe1-42b0-82de-55c143c3754a.png)

I had to this line my `_config.yml` file:

```yaml
ignore_theme_config: true
```

Now only the additional aux link appears as expected:

![image](https://user-images.githubusercontent.com/1271782/132730202-4cc8ae48-0aeb-48d6-a073-d23dc9f4d51b.png)

Everything else works great :+1:

aschrijver commented:
Sorry to be off-topic-ish. Maybe overlooking something obvious, but I can't figure out why in @pdmosses [combination-rec-nav](https://github.com/pdmosses/just-the-docs/tree/combination-rec-nav) branch the body font size is set to 12px and here in master it defaults to 16px.

mattxwang commented:
I think this (or a subset of the PRs in this PR) is a great candidate for #779. 

* @pdmosses, what do you think?
* for @pmarsceill, would you prefer to merge each PR in separately? is this too big of a change to introduce in one version?

pdmosses commented:
@mattxwang when I made this combination PR, I needed to resolve some conflicts between the included PRs. It was more than a year ago, and I don't recall exactly what those conflicts were (apologies for my inadequate commit comments) but AFAIK there are no remaining conflicts. However, there may well be better ways of resolving the original conflicts. Perhaps one of the other maintainers could check for that when reviewing? – although simply reviewing such a large collection of commits could already be somewhat daunting, I fear…

If @pmarsceill prefers us to add each included PR separately, I can't offer to do that myself at present, but I could be one of the reviewers of the result of somebody else doing it.

Moreover, @AlanFoster reported a couple of bugs in this combination  PR (see https://github.com/just-the-docs/just-the-docs/pull/578#issuecomment-916305038):

1. I could try to fix the one concerning a Liquid error that arises when there are no collections, to avoid the need for a config setting. (The tests that I've added in some PRs involve only non-empty collections; I guess that I overlooked the need to check what happens in the absence of test collections...)
2. I don't know how Jekyll decides which settings in the theme config automatically appear in the user's config, and I've no idea how to avoid the need for the workaround of ignoring the theme config, found by @AlanFoster.

pmarsceill commented:
We can totally merge this into #779 once the fixes described by @pdmosses are addressed.

> I could try to fix the one concerning a Liquid error that arises when there are no collections, to avoid the need for a config setting. (The tests that I've added in some PRs involve only non-empty collections; I guess that I overlooked the need to check what happens in the absence of test collections...)

Yes, I think this would just be doing a check to make sure that there are indeed collections before executing the liquid to do something with them.

> I don't know how Jekyll decides which settings in the theme config automatically appear in the user's config, and I've no idea how to avoid the need for the workaround of ignoring the theme config, found by @AlanFoster.

Jekyll doesn't assume any theme config by default, generally we try to bake in reasonable defaults into the execution in liquid when there is no config defined. There must have been some additional Aux links defined somewhere (maybe in the testing setup?) and by default we show all of them.

pdmosses commented:
Thanks @pmarsceill. I'll now try to get this combination PR ready for merging.

pdmosses commented:
I've discovered that the default order of pages with numeric titles and no `nav_order` field depends on the Jekyll version; I've added a warning about that to the relevant test page.

The issues noted by @AlanFoster were when using different (but closely related) PRs; further details would help with determining whether they also affect the current PR.

pdmosses commented:
## Overview

The development of this PR involved the following main steps, mostly corresponding to the referenced PRs:

1. Fix code highlighting in `dark` mode (#463).
2. Show `docs` pages also when `just-the-docs: collections` are specified, and move `docs/tests` to `_tests` (#448).
3. Add callouts (#466).
4. Fix breadcrumbs list for grandchildren whose parents have the same title (#477).
5. Avoid rake command corrupting search data (#495).
6. Fix list labels for nesting of unordered lists in ordered lists (#496).
7. Set color for search input text (#498).
8. Fix a bug for sites with no child pages.
9. Remove color setting for `h5` and `h6` (#508).
10. Fix TOC/breadcrumbs links for multiple collections (#494).
11. Add collection configuration option `nav_fold`.
12. Adjust indentation and color for folded collection navigation.
13. Scroll navigation to show the link to the current page (#639).

Some intermediate commits made various improvements and fixed minor bugs. Apologies for the difficulty of reviewing this PR with its long and messy commit history!

mattxwang commented:
I've changed the base branch to v0.4.0 now. It seems like there are no conflicts, which is stellar! I'll try to review all of the functionality in the next few days and merge this in.

pdmosses commented:
@mattxwang the PR branch still has some occurrences of `absolute_url`, which should be replaced by `relative_url` as explained in #543; see also #359.

This was deliberate: the `combination` PR combines only my own PRs, so it doesn't include #544, which should be merged separately into v0.4.0, I think. (It might be simpler to merely do a global replacement after merging this PR, but @svrooij or  @marksie1988 should preferably get the credit for the update.)

pyorot commented:
How is this going? There are some features in here and #462 that I've been waiting to use in a website I maintain for 14 months but have never made it into the main codebase. They seem to all be made but not reviewed. Is it maybe the move to merge them all into the v0.4.0 dev branch with minimal review, then work on debugging? I'm sure many users will switch over and reрort any integration bugs.

svrooij commented:
@pdmosses you can tag users in the commit by generating a special Co-author commit message. Use [this page](https://svrooij.io/github-info.html) or check [this page](https://svrooij.io/2020/07/31/add-extra-author/) for the explanation.

mattxwang commented:
> How is this going?

@pyorot sorry - I've been trying to set aside time to review this PR, but recovery from surgery has definitely been a bit challenging. I should be able to give it a formal review this week; I don't anticipate any large changes, and then it's up to @pmarsceill on our next release cycle for v0.4.0. Currently I don't have the permissions to push a new release, so that would be the blocker.

pyorot commented:
@mattxwang No worries; best wishes on recovery. I was more wondering if there's a chance to get the commits down into a рermanent branch that users can reference as a remote theme and start using and testing.

Also, I thought from [this](https://github.com/just-the-docs/just-the-docs/discussions/759#discussioncomment-2134223) that @pmarsceill was mostly inactive and wouldn't be deрended on in the release рrocess?

mattxwang commented:
So sorry - honestly still recovering from surgery and got hit with graduation. Going to get a review in today/tomorrow, with the aim of catching anything egregious.

> Also, I thought from [this](https://github.com/just-the-docs/just-the-docs/discussions/759#discussioncomment-2134223) that @pmarsceill was mostly inactive and wouldn't be deрended on in the release рrocess?

Yes, that was the plan! Perhaps @pdmosses has a different permissions set than I do, but I personally still don't have access to many repository features (force merge, branch protection rules, release publication, etc.). I also don't have access to RubyGems. I think we'll need Patrick one last time to transfer that, unless Prof Mosses has a more expanded permissions-set than me.

mattxwang commented on _tests/navigation/order/default/index.md:
```diff
@@ -13,3 +13,9 @@ When `nav_order` fields are omitted, the pages are ordered alphabetically by the
+The default order of numeric titles depends on which version of Jekyll is used
+to build the site: 
+the default build on GitHub Pages uses Jekyll3.8 and gives lexicographic order,
+but when using Jekyll4, `2` precedes `10`.
```
Nit: 
```suggestion
but when using Jekyll 4, `2` precedes `10`.
```

mattxwang commented on _tests/navigation/index.md:
```diff
@@ -0,0 +1,28 @@
+parent: Tests Home
+---
+
+# Navigation
```
A bit concerned about how brittle the links are to things moving around, but I have no strong opinions.

pdmosses replied:
I don't see why you're concerned. If the tests are moved to a separate repo without changing the file titles, the links should automatically remain valid, since the navigation hierarchy for a Just the Docs website is independent of where the files are located in its repo.

mattxwang commented on docs/configuration.md:
````diff
@@ -137,6 +137,59 @@ jtd.addEvent(toggleDarkMode, 'click', function(){
+[^dark]:
+    If you use the `dark` color scheme, this callout uses `$red-300` for the background, and `$red-000` for the title. 
+
+The colors `grey-lt`, `grey-dk`, `purple`, `blue`, `green`, `yellow`, and `red` are predefined; to use a custom color, you need to define its `000` and `300` levels in your SCSS files. For example, to use `pink`, add the following to your `_sass/custom/custom.scss` file:
+
+```scss
+$pink-000: #f77ef1;
+$pink-100: #f967f1;
+$pink-200: #e94ee1;
+$pink-300: #dd2cd4;
+```
````
With the new addition in this PR, this is implicitly done (until the user overrides `custom.scss`), right? Or am I missing something?

pdmosses replied:
Those shades of pink were defined here in [`_sass/custom/custom.scss`](https://github.com/pdmosses/just-the-docs/blob/combination/_sass/custom/custom.scss) only for use in [callouts.md](https://github.com/pdmosses/just-the-docs/blob/combination/_tests/color/callouts.md), which is located in `_tests`. When we move all the tests to a new repo, the motivation for choosing these particular shades of pink will disappear, and `_sass/custom/custom.scss` should simultaneously revert to its usual blank state. (If it didn't, we'd need to mention and motivate the contents of the file.)

mattxwang replied:
Understood, thanks!

mattxwang commented on docs/configuration.md:
````diff
@@ -137,6 +137,59 @@ jtd.addEvent(toggleDarkMode, 'click', function(){
+    opacity: 0.3
+```
+
+You can change the default opacity (0.2) for all callouts, e.g.:
````
Nit:
```suggestion
You can change the default opacity (`0.2`) for all callouts, e.g.:
```

mattxwang commented on docs/configuration.md:
```diff
@@ -149,39 +202,43 @@ ga_tracking_anonymize_ip: true # Use GDPR compliant Google Analytics settings (t
+You can also use [Jekyll collections](https://jekyllrb.com/docs/collections/) which group documents semantically together.
 
-For example, put all your documentation files in the `_docs` folder and create the `docs` collection:
+For example, put all your test files in the `_tests` folder and create the `tests` collection:
```
Just curious - is there a reason that the default is changed to `tests` instead of `docs` (and some of the default options are different)?

pdmosses replied:
Yes, I think so: having a folder named `_docs` for creating a collection in the `docs` folder seems potentially confusing and risky when one might already have *manually* created files in `docs`! What happens if *both* folders contain an (html) file with the same name? I'm not sure the Jekyll developers envisaged that use case, and it's unclear (at least to me) which file would 'win' in the resulting website. Even if it's obvious to Jekyll experts, or stated somewhere in the Jekyll docs, I can see no motivation for *encouraging* our users to name a collection folder `_docs` (especially when we also encourage them to put their source files in `docs`!).

When collections were first introduced in Just the Docs, users who wanted to use a collection needed to move all their source files for ordinary pages into collection folders. Then there were no longer any source files in `docs`, so no confusion could arise. One of the PRs incorporated in the present `combination` is #448, allowing ordinary pages and collections to coexist. I personally find that feature a useful enhancement of our support for collections, and I don't recall any objections being raised to it.

A less significant motivation was in connection with adding tests for intra-collection navigation in the current repo.

mattxwang replied:
I see, thanks for the context - much appreciated!

mattxwang commented on docs/ui-components/callouts.md:
```diff
@@ -0,0 +1,104 @@
+[^postfix]:
+    You can put the callout markup either before or after its content.
+
+#### An untitled callout
```
This breaks semantic HTML heading order. Do you think `##`/`<h2>` would be appropriate to use here?

pdmosses replied:
Similar breaks occur in the existing docs pages – e.g., see the following near the top of [navigation-structure.md](https://github.com/pdmosses/just-the-docs/blob/combination/docs/navigation-structure.md):

```markdown
## Ordering pages

To specify a page order, you can use the `nav_order` parameter in your pages' YAML front matter.

#### Example
```

I followed the same pattern for the sake of visual uniformity with the (unobtrusive) heading style used to introduce examples.

In general, however, I do prefer to keep to semantic HTML. A compromise here would be to use the same *visual* style as `<h4>` (and add an anchor to facilitate citation in discussions of examples). But ideally, that should be done uniformly for all docs pages…

I suggest that we leave the `####`s in [callouts.md](https://github.com/pdmosses/just-the-docs/blob/combination/docs/ui-components/callouts.md), and revisit this docs-only issue after v0.4.0.

mattxwang replied:
Sounds good - I'll write up an issue for this.

mattxwang replied:
Have just made #870!

mattxwang commented on docs/ui-components/callouts.md:
````diff
@@ -0,0 +1,104 @@
+#### An untitled callout
+{: .no_toc }
+
+```markdown
+{: .highlight }
+A paragraph
+```
````
What are your thoughts on including some custom styles so that users can see callouts in action on this page? Perhaps overriding some variables just on this page?

(this can be deferred until later)

pdmosses replied:
To show callouts in action would be good practice, and consistent with the illustrations on the pages for buttons, labels, etc.

However, the uniform association of callouts with particular colors throughout a website currently uses config files (in contrast to the ad hoc associations of buttons and labels with colors). I'm not sure how easy it would be to simulate such configuration on a particular page.

Let's defer it.

mattxwang replied:
Gotcha, sounds good. Backlog it is!

pdmosses replied:
Assuming that issue #171 and PR #466 will be closed when this PR is merged, I think the best will be if I submit a new issue for adding examples of the appearance of callouts to the docs. But it doesn't make sense to submit it before v0.4.0 has been released…

mattxwang commented on _layouts/default.html:
```diff
@@ -48,21 +48,41 @@
     </div>
     <nav role="navigation" aria-label="Main" id="site-nav" class="site-nav">
+      {% assign pages_top_size = site.html_pages
+          | where_exp:"item", "item.title != nil"
```
Just to check - is this a breaking change behaviour-wise?

pdmosses replied:
I regard it as a bug fix. If a website using v0.3.3 has ordinary pages as well as (Just the Docs) collections, the ordinary pages are not displayed; after this PR, all ordinary pages will be displayed, as well as the collections.

mattxwang reviewed (changes requested):
This was a bit of a doozy to review. I'll separate my bigger feedback into "merge-blocking" and "general feedback".

Merge-blocking:
1. **Can we use the copied Syntax-Themes files**? I am unsure if this is a license violation. See my in-code comment on `_sass/vendor/pygments/README.md`
2. For regression tests: I really like these; I think they do a great job of documenting finer code changes. Is the intention to keep these in the repo, or move them to a different repo? If it's the former, how does this affect our bundle size?
3. Do we need to add any documentation to the "Navigation Structure" file? I don't see anything in this PR; I do think we've made quite a few changes that we should note somewhere.

General feedback:

1. This isn't a regression necessarily introduced in this PR, but with the advent of new color palettes (like callouts) I do want to see us maintainers set some sort of color contrast standard. As of now I'm not the most confident that all the defaults pass WCAG 2.0.
2. Similarly, the sheer size of this content makes me think that we need better automated CI tests - particularly around semantic HTML, invalid generated CSS, etc. Something we can table for later.
3. Do we want to add the "new" tag on some of these features? Ex callouts or navigation changes?

And for posterity's sake, I have explicitly tested:
1. The behaviour of each test in `docs/tests` (ex navigation) and the `test_collections` behaviours 
4. Dark highlighting - nothing egregious jumps out
5. Various navigation pages
6. General clicking around
7. Callouts
8. the `rake` command - works as intended, no errors

I haven't done an in-depth diff on styling, and it's possible that I will have missed minor breaking changes. Nothing seems absolutely broken though! In addition, I can't say I'm fully grokking the new liquid navigation code. It seems as if nothing was wrong, but I think more user-testing will introduce edge-cases.

svrooij commented:
My PR (for fixing a lot of absolute urls) is two years old.

I won't be contributing to this repository because the release process is pretty complicated and takes a long time.

Maybe you guys should opt for simple feature updates. Just merging each pr as soon as possible and continue from that. I've been using my own fork with the urls fixed for some years now. I also recommend automating everything, so that there is not a lot of human tasks involved.

mattxwang commented:
> I won't be contributing to this repository because the release process is pretty complicated and takes a long time.
> 
> Maybe you guys should opt for simple feature updates. Just merging each pr as soon as possible and continue from that. I've been using my own fork with the urls fixed for some years now. I also recommend automating everything, so that there is not a lot of human tasks involved.

I'm sorry to hear that. For what it's worth, new maintainers have only been recently onboarded - for about two years, the project *has* been in limbo. I'm doing my best to triage and review everything, and yes - the goal is eventually to get to simple feature updates. Unfortunately there's a large backlog of PRs that I'm trying to selectively merge, and this one is high up on the priorities. Similar notes on automation (I have a few open PRs here as well).

Hopefully, we can get things running back up to speed and start on a regular release schedule. We'll always welcome your contributions if you change your mind!

pyorot commented:
Good work, Matt 😊

I’ll try to be around to pester with suggestions if anything slows down 😉

pdmosses commented:
@mattxwang many thanks for the review! Here's my initial response to the first section of your general feedback:

> Merge-blocking:
> 
> 1. Can we use the copied Syntax-Themes files? I am unsure if this is a license violation. See my in-code comment on _sass/vendor/pygments/README.md

Sorry, I know very little about licenses, and I can't answer that. I'd like to make the code highlighting closer to what GitHub uses, based on [Starry Night](https://github.com/wooorm/starry-night). Its license section states:

> The grammars included in this package are covered by their repositories’ respective licenses, which are permissive (apache-2.0, mit, etc), and made available in [notice](https://github.com/wooorm/starry-night/blob/main/notice).
> 
> All other files [MIT](https://github.com/wooorm/starry-night/blob/main/license) © [Titus Wormer](https://wooorm.com/)

Perhaps Just the Docs could do something similar?

> 2. For regression tests: I really like these; I think they do a great job of documenting finer code changes. Is the intention to keep these in the repo, or move them to a different repo? If it's the former, how does this affect our bundle size?

They should definitely be moved to a separate repo. That will also make it possible to display the results of running the tests, as in https://pdmosses.github.io/just-the-docs/.

> 3. Do we need to add any documentation to the "Navigation Structure" file? I don't see anything in this PR; I do think we've made quite a few changes that we should note somewhere.

I think most of the changes are bug fixes. I was aiming at backwards compatibility, modulo bug fixes. Do you have particular changes in mind?

pdmosses commented on _config.yml:
```diff
@@ -100,6 +92,61 @@ gh_edit_view_mode: "tree" # "tree" or "edit" if you want the user to jump into t
+# Examples (not provided by default):
+# 
+# callouts:
+#   admonition:
+#     # title:
+#     color: grey-lt
+#   attention:
+#     title: Attention
+#     color: purple
+#   caution:
+#     title: Caution
+#     color: yellow
+#   custom:
+#     title: Custom
+#     color: pink
+#     opacity: 0.3
+#   danger:
+#     title: Danger
+#     color: red
+#   error:
+#     title: Error
+#     color: red
+#   highlight:
+#     color: yellow
+#     opacity: 0.5
+#   hint:
+#     title: Hint
+#     color: grey-dk
+#   important:
+#     title: Important
+#     color: green
+#   note:
+#     title: Note
+#     color: blue
+#   tip:
+#     title: Tip
+#     color: grey-dk
+#   warning:
+#     title: Warning
+#     color: red
```
I think they should *all* be removed from the config! The explanations in https://github.com/pdmosses/just-the-docs/blob/combination/docs/ui-components/callouts.md and https://github.com/pdmosses/just-the-docs/blob/combination/docs/configuration.md#callouts should be adequate (or improved if they aren't, although that may have to be deferred to a later version).

Perhaps the same principle should apply to all the other configurable elements too: remove the commented-out options from `_config.yml`. The theme docs shouldn't need to refer to the config file used to produce the website, except  maybe in a (new) section aimed at advanced users who are going to base a new theme on this one.

For now, I'm removing from `_config.yml` only the cited commented-out lines regarding callouts. Pruning comments from the rest of the config file should be discussed separately.

However, note that to illustrate the effect of callout config options in the theme web pages, we'd need to add them (*not* commented-out) in the theme config.

BTW, I think the following comment at the beginning of https://just-the-docs.github.io/just-the-docs/docs/configuration/ is particularly confusing for new users:

> View this site’s [_config.yml](https://github.com/just-the-docs/just-the-docs/tree/main/_config.yml) file as an example.

New users should generally start with a much simpler config. Perhaps the new repo for Just the Docs tests could have an example of such a simple config, and the theme docs could link to that.

mattxwang replied:
Understood, thank you for the response.

> Perhaps the same principle should apply to all the other configurable elements too: remove the commented-out options from _config.yml. 

Hm, that's an interesting take. I personally dislike it when config files get too big with commented-out examples, but I'm also concerned from a usability perspective - I do know many of our users don't have significant background in web development, configs, etc.

> New users should generally start with a much simpler config. Perhaps the new repo for Just the Docs tests could have an example of such a simple config, and the theme docs could link to that.

I like the suggestion, especially wrt #756.

pdmosses replied:
From a usability perspective, I think the best for all users is to read _explanations_ and _illustrations_ of the various configuration options in the [Configuration](https://just-the-docs.github.io/just-the-docs/docs/configuration/) section of the theme docs, rather than expecting them to browse the raw `_config.yml` file, where some of the settings are not intended for general use. Not all the options are currently explained.

Moreover, I think we should consider dividing the current [Configuration](https://just-the-docs.github.io/just-the-docs/docs/configuration/) page into a page for each feature (or group of related features, such as customising the content of the different parts of pages). But that can wait to a later release.

pdmosses replied:
> But that can wait to a later release.

We could also improve the theme docs _between_ releases: a PR to fix typos or improve navigation on the theme website can be merged into `main` as soon as it has been approved, without triggering a new release.

pdmosses commented on _config_tests.yml:
```diff
@@ -0,0 +1,97 @@
+# Regression tests
```
Yes, all the tests and their config files are to be moved to a new repo for Just the Docs tests.

BTW, I have a small tests repo for Just the Docs at https://pdmosses.github.io/just-the-docs-tests/. It uses GitHub Actions to build the site on GitHub Pages with Jekyll 4. I'd be happy to transfer it to the Just the Docs organization, but it will surely be better to create a fresh tests repo with a best-practice minimal config (and avoid inheriting my messy commit history).

mattxwang replied:
Great. Do you want to do that before or after this PR is merged? I personally do have permissions to create a new repository in the organization - would you want me to do that?

(also happy to just copy over your existing jtd-tests repo, sans commit history if you'd like)

pdmosses commented on _sass/vendor/pygments/README.md:
```diff
@@ -0,0 +1,48 @@
+# Pygments
+
+> The accompanying files were copied from https://github.com/StylishThemes/Syntax-Themes/tree/master/pygments
```
Quoting my earlier response to your general comments:

> Sorry, I know very little about licenses, and I can't answer that. I'd like to make the code highlighting closer to what GitHub uses, based on [Starry Night](https://github.com/wooorm/starry-night). Its license section states:
>
> > The grammars included in this package are covered by their repositories’ respective licenses, which are permissive (apache-2.0, mit, etc), and made available in [notice](https://github.com/wooorm/starry-night/blob/main/notice).
> > 
> > All other files [MIT](https://github.com/wooorm/starry-night/blob/main/license) © [Titus Wormer](https://wooorm.com/)
> 
> Perhaps Just the Docs could do something similar?

pdmosses commented on _sass/vendor/pygments/index.html:
```diff
@@ -0,0 +1,858 @@
+<!DOCTYPE html>
```
No, I think we can remove this file. At the time, I thought it might be useful when browsing the theme docs locally, but I now think the theme website should merely link to the relevant page in the Pygments website.

pdmosses commented on _config_tests.yml:
```diff
@@ -0,0 +1,97 @@
+  test_collection_4:
+    permalink: "/:collection/:path:output_ext"
+    output: true
+  test_collection_4:
```
It was a typo. I'm correcting it.

pdmosses commented on _tests/navigation/exclude/excluded-child.md:
```diff
@@ -0,0 +1,15 @@
+---
+layout: default
+title: Excluded Child
+parent: Not Excluded
```
I'm correcting it to:

```yaml
parent: Non-excluded
```

pdmosses commented on _tests/navigation/order/default/index.md:
```diff
@@ -13,3 +13,9 @@ When `nav_order` fields are omitted, the pages are ordered alphabetically by the
+{: .note }
+The default order of numeric titles depends on which version of Jekyll is used
+to build the site: 
+the default build on GitHub Pages uses Jekyll3.8 and gives lexicographic order,
```
I'll update that.

Perhaps some day in the not-too-distant future, we'll stop using Jekyll 3.8.5 for building sites on GitHub Pages, and allow users to specify the required version of Jekyll in the config file?

Users new to Jekyll presumably install the latest version of Jekyll to browse their site locally. I think it's really confusing that they can't use all the latest Jekyll features when building on GitHub Pages (without learning first how to set up GitHub Actions) – especially as the [Jekyll docs pages](https://jekyllrb.com/docs/) aren't versioned…

pdmosses commented:
@mattxwang, I think I've now responded to all your detailed comments on this combined PR; let me know if I've overlooked any. 

I've also remembered to replace `absolute_url` by `relative_url` in all files.

> General feedback:
> 
> 1. This isn't a regression necessarily introduced in this PR, but with the advent of new color palettes (like callouts) 

The callouts implementation does not choose *any* colors at all. The (new) configuration section on callouts states "To use this feature, you need to configure a `color` and (optionally) `title` for each kind of callout you want to use".

> I do want to see us maintainers set some sort of color contrast standard.

Yes. We'd had complaints about the contrasts in default code highlighting (especially) in our dark color scheme. An improvement for dark code highlighting is already included in this PR, but I don't know how accessible it is to the visually impaired. And I don't know whether @pmarsceill made our default light code highlighting based on some previous scheme, or designed it himself.

As I previously mentioned, I now think the best way forward would be to use the same code highlighting as GitHub's PrettyLights. But that can wait to some future release.

Buttons and labels have palettes of pre-defined color variables; I don't know whether they are based on some standard.

> As of now I'm not the most confident that all the defaults pass WCAG 2.0.

Is there some way of checking?

> 2. Similarly, the sheer size of this content makes me think that we need better automated CI tests - particularly around semantic HTML, invalid generated CSS, etc. Something we can table for later.

A few years ago, I tried using the [W3C validator](https://validator.w3.org), which found some issues. It would be good to include it in our build checks.

> 3. Do we want to add the "new" tag on some of these features? Ex callouts or navigation changes?

I'd prefer to mention how new, e.g., "New in v0.4" on callouts. For changes to previous features such as navigation, I suggest to add "Changed in v0.4" on the changed sections.

In both cases, the label could link to the PR that implements the new or changed feature.

> And for posterity's sake, I have explicitly tested:
> 
> 1. The behaviour of each test in docs/tests (ex navigation) and the test_collections behaviours
> 2. Dark highlighting - nothing egregious jumps out
> 3. Various navigation pages
> 4. General clicking around
> 5. Callouts
> 6. the rake command - works as intended, no errors

Great, thanks!

> I haven't done an in-depth diff on styling, and it's possible that I will have missed minor breaking changes. Nothing seems absolutely broken though! 

Bug fixes (to navigation, layout, color, etc.) are deliberately 'breaking' the v0.3.3 behavior.

It would be good to be able to quickly release a v0.4.1 update to fix any significant issues arising with v0.4.0. But it's easy enough to configure a website to continue using v0.3.3; the docs should explain how.

> In addition, I can't say I'm fully grokking the new liquid navigation code.

If you let me know which bits need (further) explanation, I can try and add/improve the comments.

For [multi-level navigation](https://github.com/pdmosses/just-the-docs/tree/rec-nav-2) (which I'm using myself for a moderate-sized website), I've written fairly detailed [documentation](https://pdmosses.github.io/just-the-docs/tests/navigation/recursion/about.html). The 3-level navigation in the present PR is based on the same implementation principles.

> It seems as if nothing was wrong, but I think more user-testing will introduce edge-cases.

I've tried to make the current navigation (and other) tests reasonably rigorous. If new edge-cases are reported, we can add them to the tests for the subsequent release. By moving the tests to a separate repo, we can add tests independently of new theme releases.

mattxwang commented:
Thanks for the response - I know this is a lot!

In general, I agree with almost everything you've written! I think there are only two (or possibly just one) blockers:

1. moving the tests to a test repo - I'll defer to your judgement on when that needs to happen (though from my perspective, before this PR seems to make more sense)
2. licensing - see below

## Licensing

Upon further reading, I'm pretty certain that we cannot use this repo ([Syntax-Themes](https://github.com/StylishThemes/Syntax-Themes)) in just-the-docs, since we don't follow the ShareAlike clause of CC-BY-SA 4.0. If this was just CC-BY, we'd be fine. I believe this is the licensing issue alluded to in starry-night.

I want to minimize the amount of work you have to do, though I'm not particularly familiar with the syntax theme landscape. I am looking at [starry-night's notice](https://github.com/wooorm/starry-night/blob/main/notice) and it seems like all the CSS files there are permissively released. Are you able to directly use `starry-night`'s CSS?

Alternatively, it seems like many of the source files are taken from Atom. There seem to be Jekyll-packaged Pygments/Rouge-compatible Atom-based stylesheets, like this one: https://github.com/mgyongyosi/OneDarkJekyll . Any thoughts on that?

pdmosses commented:
> In general, I agree with almost everything you've written! 

@mattxwang Thanks for resolving all those points!

> I think there are only two (or possibly just one) blockers:
> 
> 1. moving the tests to a test repo - I'll defer to your judgement on when that needs to happen (though from my perspective, before this PR seems to make more sense)

I don't think the config in my just-the-docs-tests repo is minimal, so best to create a fresh site – perhaps following our Quick Start advice on the home page?

I'll then add extra config files for testing the different features. That will allow starting from tests in the absence of collections (I recall fixing a bug due to testing nav only when collections are present).

I definitely think it's best to move all those test files out of the theme website before merging this PR, as it will greatly reduce the number of files that it adds/changes.

> 2. licensing - see below

pdmosses commented:
(Sorry, I clicked the wrong button and closed this PR by mistake…)

> Are you able to directly use starry-night's CSS?

Not unless the class names used by Pygments/Rouge can be directly mapped to the `.pl-*` class names used by starry-night's CSS.

> Alternatively, it seems like many of the source files are taken from Atom. There seem to be Jekyll-packaged Pygments/Rouge-compatible Atom-based stylesheets, like this one: https://github.com/mgyongyosi/OneDarkJekyll . Any thoughts on that?

Thanks for the pointer to that possibility. I'll take a look, and try to find one that matches GitHub's colours.

mattxwang commented:
Sounds good. I can get to making the test repo another day. Do you need help with the syntax highlighting, or would you prefer to take point on it?

pdmosses commented:
> I can get to making the test repo another day. 

OK. I'll now copy the current branch used for this PR to a new branch, then remove the tests from the PR.

> Do you need help with the syntax highlighting, or would you prefer to take point on it?

I'm currently stuck: the [GitHub syntax theme for Atom](https://atom.io/themes/atom-github-syntax) has great accessibility features, but it is based on the TextMate CSS classes used by GitHub (`pl-*`, listed [here](https://github.com/Alhadis/language-etc/blob/master/samples/lists/scope-previews.nanorc)), and we need to use Rouge/Pygments classes in our Jekyll theme.

I may well have overlooked something when Googling: the word 'GitHub' matches all projects…

pdmosses commented:
@mattxwang the SCSS provided by [jekyllHighlighterAtomOneDarkTheme](https://github.com/bryanchapel/jekyllHighlighterAtomOneDarkTheme) might work for our dark scheme. Would the license work too?

mattxwang commented:
@pdmosses we wouldn't be able to use [jekyllHighlighterAtomOneDarkTheme](https://github.com/bryanchapel/jekyllHighlighterAtomOneDarkTheme), since it has a GPL (copyleft) license. The latter, [OneDarkJekyll](https://github.com/mgyongyosi/OneDarkJekyll), is completely good!

pdmosses commented:
> The latter, [OneDarkJekyll](https://github.com/mgyongyosi/OneDarkJekyll), is completely good!

Thanks for the confirmation. 

As we can't include the current Pygments code for _Tomorrow Night_ in our releases, I might as well replace it by OneDarkJekyll in the current PR (after checking that it's at least an improvement on the code highlighting in v0.3.3). I'll also look for a comparable light mode.

BTW, I'm wondering whether it's possible to use (a link to) [highlight.js](https://highlightjs.org/) on Jekyll sites? The quality of their GitHub highlighting options on the [demo site](https://highlightjs.org/static/demo/) is impressive!

pdmosses commented:
> Sorry, I'm a bit confused - what do you mean by "a link to"? Like embedding it with the `<script>` tag?

Yes – the question is whether Jekyll can use highlight.js instead of Rouge. I'm not familiar with the various kramdown options. After experimenting with OneDarkJekyll, I'll take a look at that.

pdmosses commented:
@mattxwang I don't think the _JekyllOneDark_ code highlighting is particularly good (and I doubt it's WCAG-compliant, but I don't have expertise on that issue).

I have found some Pygments styles with WCAG AA and free to use: [pygments-high-contrast-stylesheets](https://github.com/mpchadwick/pygments-high-contrast-stylesheets). Could you take a look and let me know which we might use instead of _Tomorrow Night_? As we're unsure about the provenance and WCAG rating of our current light code highlighting, we could change that too.

Jekyll can use highlight.js (by disabling the highlighter in the kramdown options) and it has impressive styles for both light and dark mode (including GitHub-like). However, I think it might be too big a risk to suddenly drop Rouge and switch to highlight.js in v0.4.0. Do you agree, or would you like me to continue testing it? I could also submit a separate PR for it.

pdmosses commented:
@mattxwang I hope the replacement of Pygments by OneDarkJekyll has resolved the merge-blocking license issue.

I'll leave further experimentation with highlighting dark code (aiming at WCAG 2 compliance) to a separate PR.

mattxwang reviewed (approved):
Hi @pdmosses, just getting back to everything now. This all looks good, and thanks for making the theme swap so quickly. I think we're good to go re: merging.

> I have found some Pygments styles with WCAG AA and free to use: [pygments-high-contrast-stylesheets](https://github.com/mpchadwick/pygments-high-contrast-stylesheets). Could you take a look and let me know which we might use instead of Tomorrow Night? As we're unsure about the provenance and WCAG rating of our current light code highlighting, we could change that too.

Apologies for not getting back to you on the Pygments styling earlier. FWIW, I think monokai is a good dark theme to pick that is also widely used by many people in the community.

> Jekyll can use highlight.js (by disabling the highlighter in the kramdown options) and it has impressive styles for both light and dark mode (including GitHub-like). However, I think it might be too big a risk to suddenly drop Rouge and switch to highlight.js in v0.4.0. Do you agree, or would you like me to continue testing it? I could also submit a separate PR for it.

I think I agree that this is too big of a change for v0.4.0. However, I think experimenting with it is an interesting idea. I personally use Jekyll because I like shifting as much work on build (versus on page load, which is what `highlight.js` would need for the base out-of-the-box use-case), but soliciting more community input would be good.
